engine/apps/alerts/tasks/notify_user.py

```diff
@@ -418,17 +487,119 @@ def perform_notification(log_record_pk):
+            entry["notification_policy_id"] = (
+                entry["notification_policy_id"]
+                if entry["notification_policy_id"] in existing_notification_policies
+                else None
+            )
+            # collect notification data for active alert groups
+            current_notification_data.append(entry)
+
+            log_record = UserNotificationPolicyLogRecord(
+                author=user_notification_bundle.user,
+                type=UserNotificationPolicyLogRecord.TYPE_PERSONAL_NOTIFICATION_TRIGGERED,
+                notification_policy_id=entry["notification_policy_id"],
+                alert_group_id=entry["alert_group_id"],
+                slack_prevent_posting=True,
+                notification_step=UserNotificationPolicy.Step.NOTIFY,
+                notification_channel=user_notification_bundle.notification_channel,
+            )
+            log_record.save()
+            created_lod_records.append(log_record)
+
+        user_notification_bundle.notification_data = []
+        user_notification_bundle.notification_task_id = None
+        user_notification_bundle.last_notified = timezone.now()
+        user_notification_bundle.eta = None
+        user_notification_bundle.save(
+            update_fields=["notification_data", "notification_task_id", "last_notified", "eta"]
+        )
 
-    from apps.base.models import UserNotificationPolicyLogRecord
+    if not user_notification_bundle.user.is_notification_allowed:
+        for notification_data in current_notification_data:
+            UserNotificationPolicyLogRecord(
+                author=user_notification_bundle.user,
+                type=UserNotificationPolicyLogRecord.TYPE_PERSONAL_NOTIFICATION_FAILED,
+                reason="notification is not allowed for user",
+                alert_group=notification_data["alert_group_id"],
+                notification_error_code=UserNotificationPolicyLogRecord.ERROR_NOTIFICATION_FORBIDDEN,
+            ).save()
+        return
+
+    if not current_notification_data:
+        task_logger.info("no alert groups to notify about, skip notification")
 
-    task_logger.debug(f"LOG RECORD PK: {log_record_pk}")
-    task_logger.debug(f"LOG RECORD LAST: {UserNotificationPolicyLogRecord.objects.last()}")
+    elif len(current_notification_data) == 1:
+        # perform regular notification
+        log_record = created_lod_records[0]
+        task_logger.info(
+            f"there is only one alert group in bundled notification, perform regular notification. "
+            f"alert_group {log_record.alert_group_id}"
+        )
+        transaction.on_commit(partial(create_perform_notification_task, log_record.pk, log_record.alert_group_id))
+    else:
+        task_logger.info(f"perform notification for alert groups with ids: {active_alert_groups}")
+        if user_notification_bundle.notification_channel == UserNotificationPolicy.NotificationChannel.SMS:
```

Comment:
Will the sms send happen as part of this task or a separate task? I'm curious to see how it works in terms of if that needs to be completed successfully before this transaction commits and clears the notification_data.

Reply:
I think it will be done in a separate task using data collected in `current_notification_data`. In this case we don't need to rework retry logic.
I was thinking about making the second part of the current task similar to the `perform_notification` task, but calling methods of notification channels async